libbpf-tools/memleak.bpf.c

```diff
@@ -0,0 +1,392 @@
+	__sync_fetch_and_sub(&existing_cinfo->bits, decremental_cinfo.bits);
+}
+
+static __always_inline int gen_alloc_enter(size_t size)
+{
+	if (size < min_size || size > max_size) {
+		return 0;
+	}
+
+	if (sample_rate > 1) {
+		if (bpf_ktime_get_ns() % sample_rate != 0) {
+			return 0;
+		}
+	}
+
+	const pid_t pid = bpf_get_current_pid_tgid() >> 32;
+	const u64 size64 = size;
+	bpf_map_update_elem(&sizes, &pid, &size64, BPF_ANY);
+
+	if (trace_all)
+		bpf_printk("alloc entered, size = %lu\\n", size64);
+
+	return 0;
+}
+
+static __always_inline int gen_alloc_exit2(void *ctx, u64 address) {
+	const pid_t pid = bpf_get_current_pid_tgid() >> 32;
+	struct alloc_info info;
+
+	const u64* size64 = bpf_map_lookup_elem(&sizes, &pid);
+	if (!size64)
+		return 0; // missed alloc entry
+
+	__builtin_memset(&info, 0, sizeof(info));
+
+	info.size = *size64;
+	bpf_map_delete_elem(&sizes, &pid);
+
+	if (address != 0) {
+		info.timestamp_ns = bpf_ktime_get_ns();
+
+		info.stack_id = bpf_get_stackid(ctx, &stack_traces, stack_flags);
+
+		bpf_map_update_elem(&allocs, &address, &info, BPF_ANY);
+
+		update_statistics_add(info.stack_id, info.size);
+	}
+
+	if (trace_all) {
+		bpf_printk("alloc exited, size = %lu, result = %lx\\n",
+				info.size, address);
+	}
+
+	return 0;
+}
+
+static __always_inline int gen_alloc_exit(struct pt_regs *ctx) {
+	return gen_alloc_exit2(ctx, PT_REGS_RC(ctx));
+}
+
+static __always_inline int gen_free_enter(const void *address) {
+	const u64 addr = (u64)address;
+
+	const struct alloc_info *info = bpf_map_lookup_elem(&allocs, &addr);
+	if (!info)
+		return 0;
+
+	bpf_map_delete_elem(&allocs, &addr);
+	update_statistics_del(info->stack_id, info->size);
+
+	if (trace_all) {
+		bpf_printk("free entered, address = %lx, size = %lu\\n",
+				address, info->size);
+	}
+
+	return 0;
+}
+
+SEC("uprobe")
+int BPF_KPROBE(malloc_enter, size_t size)
+{
+	return gen_alloc_enter(size);
+}
+
+SEC("uretprobe")
+int BPF_KRETPROBE(malloc_exit)
+{
+	return gen_alloc_exit(ctx);
+}
+
+SEC("uprobe")
+int BPF_KPROBE(free_enter, void *address)
+{
+	return gen_free_enter(address);
+}
+
+SEC("uprobe")
+int BPF_KPROBE(calloc_enter, size_t nmemb, size_t size)
+{
+	return gen_alloc_enter(nmemb * size);
+}
+
+SEC("uretprobe")
+int BPF_KRETPROBE(calloc_exit)
+{
+	return gen_alloc_exit(ctx);
+}
+
+SEC("uprobe")
+int BPF_KPROBE(realloc_enter, void *ptr, size_t size)
+{
+	gen_free_enter(ptr);
+
+	return gen_alloc_enter(size);
+}
+
+SEC("uretprobe")
+int BPF_KRETPROBE(realloc_exit)
+{
+	return gen_alloc_exit(ctx);
+}
+
+SEC("uprobe")
+int BPF_KPROBE(mmap_enter)
+{
+	size_t size = (size_t)PT_REGS_PARM2(ctx);
+
+	return gen_alloc_enter(size);
+}
+
+SEC("uretprobe")
+int BPF_KRETPROBE(mmap_exit)
+{
+	return gen_alloc_exit(ctx);
+}
+
+SEC("uprobe")
+int BPF_KPROBE(munmap_enter, void *address) {
+	return gen_free_enter(address);
+}
+
+SEC("uprobe")
+int BPF_KPROBE(posix_memalign_enter, void **memptr, size_t alignment, size_t size)
+{
+	const u64 memptr64 = (u64)(size_t)memptr;
```

Comment:
this probably won't work on 32-bit architectures properly. size_t is BPF architecture size, so always 64-bit. To support 32-bit arches properly we might need to use a parameter passed through global var that would specify pointer size